scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/resources/StateQueryParam.java

```diff
@@ -1,28 +1,65 @@
  */
 public class StateQueryParam implements VariableArgumentComputer {
 
-  public final QueriableState state;
+  public final String resourceName;
   public final TimeExpression timeExpr;
 
-  public StateQueryParam(QueriableState state, TimeExpression timeExpression) {
-    this.state = state;
+  public StateQueryParam(String resourceName, TimeExpression timeExpression) {
+    this.resourceName = resourceName;
     this.timeExpr = timeExpression;
   }
 
-  public SerializedValue getValue(Plan plan, Window win) {
-    var time = timeExpr.computeTime(plan, win);
+  public SerializedValue getValue(SimulationResults simulationResults, Plan plan, Window win) {
+    var time = timeExpr.computeTime(simulationResults, plan, win);
     if (!time.isSingleton()) {
       throw new RuntimeException(" Time expression in StateQueryParam case must be singleton");
     }
-    return state.getValueAtTime(time.start);  }
+    final var queryT = time.start;
+
+    //TODO: unify necessary generic profile operations in Profile interface to avoid special casing
+    if (simulationResults.realProfiles.containsKey(this.resourceName)) {
+      //TODO: improve the profile data structure to allow fast time-keyed query
+      //TODO: improve Window to allow querying containment in interval directly
+      //for now we look for the last matching profile segment, if any (last to get latest if any overlaps)
+      final var profile = simulationResults.realProfiles.get(this.resourceName);
+      final Predicate<LinearProfilePiece> containsQueryTimeP = (piece)
+          -> !Window.intersect(piece.window, Window.at(queryT)).isEmpty();
+      final var piece = Lists.reverse(profile.profilePieces)
+                             .stream()
+                             .filter(containsQueryTimeP)
+                             .findFirst()
+                             .orElseThrow(() -> new Error(
+                                 "Linear profile for %s not have a segment at the desired time %s"
+                                     .formatted(resourceName, queryT)));
+      return SerializedValue.of(piece.valueAt(queryT));
+    } else if (simulationResults.discreteProfiles.containsKey(this.resourceName)) {
+      //TODO: improve the profile data structure to allow fast time-keyed query
+      //TODO: improve Window to allow querying containment in interval directly
+      //for now we look for the last matching profile segment, if any (last to get latest if any overlaps)
+      final var profile = simulationResults.discreteProfiles.get(this.resourceName);
+      final Predicate<DiscreteProfilePiece> containsQueryTimeP = (piece)
+          -> !Window.intersect(piece.window, Window.at(queryT)).isEmpty();
+      final var matchPiece = Lists.reverse(profile.profilePieces).stream()
+                                  .filter(containsQueryTimeP).findFirst();
+      return matchPiece.map(piece -> piece.value).orElse(null);
+    } else {
+      return null;
```

Comment:
I think in these two cases as well we'll want to throw some sort of error or exception. Case 1: There was no segment at the desired time, and Case 2: the requested resource does not exist.